src/test/java/org/jenkinsci/plugins/workflow/cps/CpsVmExecutorServiceTest.java

```diff
@@ -24,22 +24,97 @@
+                WorkflowRun b = r.buildAndAssertSuccess(p);
+                r.assertLogContains("no problem got 3", b);
+                r.assertLogNotContains("expected to call", b);
+                return null;
+            });
+            errors.checkSucceeds(() -> {
+                p.setDefinition(new CpsFlowDefinition("class C {@Override String toString() {'never used'}}; def gstring = /embedding ${new C()}/; echo(/oops got $gstring/)", true));
+                WorkflowRun b = r.assertBuildStatus(Result.FAILURE, p.scheduleBuild2(0)); // JENKINS-27306: No such constructor found: new org.codehaus.groovy.runtime.GStringImpl java.lang.String java.lang.String[]
+                r.assertLogContains(CpsVmExecutorService.mismatchMessage("java.lang.Class", "asType", "C", "toString"), b);
+                return null;
+            });
+            errors.checkSucceeds(() -> {
+                p.setDefinition(new CpsFlowDefinition("echo(/see what ${-> 'this'} does/)", true));
+                WorkflowRun b = r.buildAndAssertSuccess(p);
+                r.assertLogContains(CpsVmExecutorService.mismatchMessage("WorkflowScript", "echo", "org.jenkinsci.plugins.workflow.cps.CpsClosure2", "call"), b);
+                r.assertLogNotContains("see what", b);
+                return null;
+            });
+            errors.checkSucceeds(() -> {
+                p.setDefinition(new CpsFlowDefinition("node {echo(/see what ${-> 'this'} does/)}", true));
+                WorkflowRun b = r.buildAndAssertSuccess(p);
+                r.assertLogContains(CpsVmExecutorService.mismatchMessage("WorkflowScript", "echo", "org.jenkinsci.plugins.workflow.cps.CpsClosure2", "call"), b);
+                r.assertLogNotContains("see what", b);
+                return null;
+            });
+            errors.checkSucceeds(() -> {
+                p.setDefinition(new CpsFlowDefinition(
+                    "@NonCPS def shouldBomb() {\n" +
+                    "  def text = ''\n" +
+                    "  ['a', 'b', 'c'].each {it -> writeFile file: it, text: it; text += it}\n" +
+                    "  text\n" +
+                    "}\n" +
+                    "node {\n" +
+                    "  echo shouldBomb()\n" +
+                    "}\n", true));
+                r.assertLogContains(CpsVmExecutorService.mismatchMessage("WorkflowScript", "shouldBomb", null, "writeFile"), r.buildAndAssertSuccess(p));
+                return null;
+            });
+            errors.checkSucceeds(() -> {
+                p.setDefinition(new CpsFlowDefinition("@NonCPS def bad() {polygon(17) {}}; bad()", true));
+                WorkflowRun b = r.buildAndAssertSuccess(p);
+                r.assertLogContains("wrapping in a 17-gon", b);
+                r.assertLogContains(CpsVmExecutorService.mismatchMessage("WorkflowScript", "bad", null, "polygon"), b);
+                return null;
+            });
+        } finally {
```

Comment:
I was hoping that a warning would be printed when CPS-transformed code is used inside a constructor, for example in something like this:

```
errors.checkSucceeds(() -> {
    p.setDefinition(new CpsFlowDefinition("class Foo { Foo(script) { script.sleep(1) } }; new Foo(this)", true));
    WorkflowRun b = r.assertBuildStatus(Result.FAILURE, p.scheduleBuild2(0));
    r.assertLogContains(CpsVmExecutorService.mismatchMessage("Foo", "<init>", null, "sleep"), b);
    return null;
});
```

For now though, that code appears to just fail directly with `hudson.remoting.ProxyException: CpsCallableInvocation{methodName=sleep, call=com.cloudbees.groovy.cps.impl.CpsFunction@3d58963, receiver=null, arguments=[org.jenkinsci.plugins.workflow.cps.DSL$ThreadTaskImpl@19cb8536]}` and no additional message is printed.

I think this is because [FunctionCallBlock.dispatchOrArg](https://github.com/cloudbees/groovy-cps/blob/1d00d440688d6320f81c57c00e5320e63fa05727/lib/src/main/java/com/cloudbees/groovy/cps/impl/FunctionCallBlock.java#L96) handles constructor invocation directly so none of the code added downstream in `ContinuationGroup.methodCall` applies. Would it make sense to add similar logic [here](https://github.com/cloudbees/groovy-cps/blob/1d00d440688d6320f81c57c00e5320e63fa05727/lib/src/main/java/com/cloudbees/groovy/cps/impl/FunctionCallBlock.java#L98) to give more context to this kind of error?

Reply:
This patch to `FunctionCallBlock` over in groovy-cps along with the patch to `CpsVmExecutorService.className` makes the test work, although I don't know if it might cause some false positives:

```
diff --git a/lib/src/main/java/com/cloudbees/groovy/cps/impl/FunctionCallBlock.java b/lib/src/main/java/com/cloudbees/groovy/cps/impl/FunctionCallBlock.java
index 94f71e2..03e4c32 100644
--- a/lib/src/main/java/com/cloudbees/groovy/cps/impl/FunctionCallBlock.java
+++ b/lib/src/main/java/com/cloudbees/groovy/cps/impl/FunctionCallBlock.java
@@ -94,6 +94,10 @@ public class FunctionCallBlock extends CallSiteBlockSupport {
                     Object v;
                     try {
                         v = e.getInvoker().contextualize(FunctionCallBlock.this).constructorCall((Class)lhs,args);
+                    } catch (CpsCallableInvocation inv) {
+                        inv.checkMismatch(lhs, Arrays.asList("<init>"));
+                        Throwable t = new IllegalStateException("Calling CPS-transformed methods inside a constructor is not supported.");
+                        return throwException(e, t, loc, new ReferenceStackTrace());
                     } catch (Throwable t) {
                         return throwException(e, t, loc, new ReferenceStackTrace());
                     }
```